packages/message-relayer/src/relay-tx.ts

```diff
@@ -0,0 +1,397 @@
+  if (sentMessageEvents.length === 0) {
+    return null
+  }
+
+  // Not sure exactly how to handle the special case that more than one message was sent in the
+  // same transaction. For now throwing an error seems fine. We can always deal with this later.
+  if (sentMessageEvents.length > 1) {
+    throw new Error(
+      `can currently only support one message per transaction, found ${sentMessageEvents}`
+    )
+  }
+
+  // Decode the message and turn it into a nicer struct.
+  const encodedMessage = sentMessageEvents[0].args.message
+  const decodedMessage = l2CrossDomainMessenger.interface.decodeFunctionData(
+    'relayMessage',
+    encodedMessage
+  )
+
+  return {
+    target: decodedMessage._target,
+    sender: decodedMessage._sender,
+    message: decodedMessage._message,
+    messageNonce: decodedMessage._messageNonce.toNumber(),
+  }
+}
+
+/**
+ * Encodes a cross domain message.
+ * @param message Message to encode.
+ * @returns Encoded message.
+ */
+const encodeCrossDomainMessage = (message: CrossDomainMessage): string => {
+  return getContractInterface(
+    'OVM_L2CrossDomainMessenger'
+  ).encodeFunctionData('relayMessage', [
+    message.target,
+    message.sender,
+    message.message,
+    message.messageNonce,
+  ])
+}
+
+/**
+ * Finds the StateBatchAppended event associated with a given L2 transaction.
+ * @param l1RpcProvider L1 RPC provider.
+ * @param l1StateCommitmentChainAddress Address of the L1StateCommitmentChain.
+ * @param l2TransactionIndex Index of the L2 transaction to find a StateBatchAppended event for.
+ * @returns StateBatchAppended event for the given transaction or null if no such event exists.
+ */
+export const getStateBatchAppendedEventByTransactionIndex = async (
+  l1RpcProvider: ethers.providers.JsonRpcProvider,
+  l1StateCommitmentChainAddress: string,
+  l2TransactionIndex: number
+): Promise<ethers.Event | null> => {
+  const l1StateCommitmentChain = new ethers.Contract(
+    l1StateCommitmentChainAddress,
+    getContractInterface('OVM_StateCommitmentChain'),
+    l1RpcProvider
+  )
+
+  const getStateBatchAppendedEventByBatchIndex = async (
+    index: number
+  ): Promise<ethers.Event | null> => {
+    const eventQueryResult = await l1StateCommitmentChain.queryFilter(
+      l1StateCommitmentChain.filters.StateBatchAppended(index)
+    )
+    if (eventQueryResult.length === 0) {
+      return null
+    } else {
+      return eventQueryResult[0]
+    }
+  }
+
+  const isEventHi = (event: ethers.Event, index: number) => {
+    const prevTotalElements = event.args._prevTotalElements.toNumber()
+    return index < prevTotalElements
+  }
+
+  const isEventLo = (event: ethers.Event, index: number) => {
+    const prevTotalElements = event.args._prevTotalElements.toNumber()
+    const batchSize = event.args._batchSize.toNumber()
+    return index >= prevTotalElements + batchSize
+  }
+
+  const totalBatches = (
+    await l1StateCommitmentChain.getTotalBatches()
```

Comment:
Can this return `null`? Nit: nested await, can just use bignumber math

Reply:
Nope, this will just return `0` by default.